### PR TITLE
Provision for authentication filter. Ref: Last mile changes

### PR DESCRIPTION
**Pull request by abhishekjain88**
There is a corresponding PR in internal repository as well.

@gittify @bargenilesh @iizrailevsky @longdogz @mooxter @coleenHolden @sudheerkethireddy Requesting review.

### CODE REVIEW

File: modules/api/pom.xml [hunk 1]
```diff
@@ -81,7 +81,6 @@
         <dependency>
             <groupId>com.intuit.autumn</groupId>
             <artifactId>autumn-api</artifactId>
-            <version>1.0.20160714060618</version>
```
Comment (iizrailevsky) [inline modules/api/pom.xml]: Thanks for pulling these versions into the central pom.xml! 👍

File: modules/authentication/src/main/java/com/intuit/wasabi/authentication/AuthenticateByHttpRequest.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ *******************************************************************************/
+
+package com.intuit.wasabi.authentication;
+
+import com.sun.jersey.api.core.HttpRequestContext;
+import java.util.Map;
+
+public interface AuthenticateByHttpRequest {
+
+    Map<String, String> authenticate(HttpRequestContext httpRequest);
```
Comment (iizrailevsky) [inline modules/authentication/src/main/java/com/intuit/wasabi/authentication/AuthenticateByHttpRequest.java]: @abhishekjain88 Mind adding a method Javadoc header to clarify what the method does & what the input/output parameters are?

File: modules/main/pom.xml [hunk 3]
```diff
@@ -43,8 +43,8 @@
 
         <application.jersey.provider.paths>com.intuit.wasabi.api;com.codahale.metrics.jersey</application.jersey.provider.paths>
         <application.jersey.pojo.enabled>true</application.jersey.pojo.enabled>
-        <application.jersey.request.filters>com.sun.jersey.api.client.filter.LoggingFilter</application.jersey.request.filters>
```
Comment (iizrailevsky) [inline modules/main/pom.xml]: @abhishekjain88 Let's make sure we test that there's no loss of functionality due to removing this filter in both requests & responses: com.sun.jersey.api.client.filter.LoggingFilter

File: modules/authentication/src/main/java/com/intuit/wasabi/authentication/impl/NoOpAuthenticateByHttpRequestImpl.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import com.sun.jersey.api.core.HttpRequestContext;
+
+import java.util.Collections;
+import java.util.Map;
+
+public class NoOpAuthenticateByHttpRequestImpl implements AuthenticateByHttpRequest {
+
+    @Override
+    public Map<String, String> authenticate(HttpRequestContext httpRequestContext) {
+        return Collections.emptyMap();
```
Comment (iizrailevsky) [inline modules/authentication/src/main/java/com/intuit/wasabi/authentication/impl/NoOpAuthenticateByHttpRequestImpl.java]: @abhishekjain88 Should we add a dummy unit test class for this to retain code coverage & overall testability? I know this is a nit. 😄

File: modules/api/pom.xml (hunk 1 above)
Comment (abhishekjain88) [inline modules/api/pom.xml]: Yes @iizrailevsky ! Need to clean this in other poms as well. But it will be better to do it as a separate change as it is unrelated.

File: modules/authentication/src/main/java/com/intuit/wasabi/authentication/AuthenticateByHttpRequest.java (hunk 2 above)
Comment (abhishekjain88) [inline modules/authentication/src/main/java/com/intuit/wasabi/authentication/AuthenticateByHttpRequest.java]: ack @iizrailevsky! Will do.

File: modules/main/pom.xml (hunk 3 above)
Comment (abhishekjain88) [inline modules/main/pom.xml]: @iizrailevsky I tested it. Existing integration tests & unit tests work fine + no change in logging of messages. Actually, it was throwing a warning message earlier and ignoring it, when it was present - so, it was really not serving any purpose. :)

File: modules/authentication/src/main/java/com/intuit/wasabi/authentication/impl/NoOpAuthenticateByHttpRequestImpl.java (hunk 4 above)
Comment (abhishekjain88) [inline modules/authentication/src/main/java/com/intuit/wasabi/authentication/impl/NoOpAuthenticateByHttpRequestImpl.java]: No problem at all @iizrailevsky . Will add it.

File: modules/main/pom.xml (hunk 3 above)
Comment (iizrailevsky) [inline modules/main/pom.xml]: Ack - let's get rid of it then! 👍

File: modules/authentication/src/main/java/com/intuit/wasabi/authentication/AuthenticateByHttpRequest.java (hunk 2 above)
Comment (abhishekjain88) [inline modules/authentication/src/main/java/com/intuit/wasabi/authentication/AuthenticateByHttpRequest.java]: Done @iizrailevsky

File: modules/authentication/src/main/java/com/intuit/wasabi/authentication/impl/NoOpAuthenticateByHttpRequestImpl.java (hunk 4 above)
Comment (abhishekjain88) [inline modules/authentication/src/main/java/com/intuit/wasabi/authentication/impl/NoOpAuthenticateByHttpRequestImpl.java]: Done @iizrailevsky !

**Comment by abhishekjain88** [comment]
@iizrailevsky Thanks for doing an extensive review of this. Appreciate it! Have tried to address your concerns. Request you to take a second look and let me know if anything needs any further change.

File: modules/authentication/src/main/java/com/intuit/wasabi/authentication/AuthenticateByHttpRequest.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+public interface AuthenticateByHttpRequest {
+
+    /**
+     * Looks for required headers in the http request and uses that information to authenticate the request.
+     * @param httpRequest
+     * @return a map of any additional information obtained during the authentication process
+     * which might be useful in subsequent code flow.
+     */
+    Map<String, String> authenticate(HttpRequestContext httpRequest);
```
Comment (gittify) [inline modules/authentication/src/main/java/com/intuit/wasabi/authentication/AuthenticateByHttpRequest.java]: The return type needs to be Map<String, List<String>> based on our previous discussion (we are using this value to overwrite auth header)

Comment (abhishekjain88) [inline modules/authentication/src/main/java/com/intuit/wasabi/authentication/AuthenticateByHttpRequest.java]: ok @gittify Let me instead make it as a generic so that it can support future use cases as well.

**Comment by bargenilesh** [comment]
@abhishekjain88 Changes looks good :shipit:

File: modules/authentication/src/main/java/com/intuit/wasabi/authentication/AuthenticateByHttpRequest.java (hunk 5 above)
Comment (abhishekjain88) [inline modules/authentication/src/main/java/com/intuit/wasabi/authentication/AuthenticateByHttpRequest.java]: @gittify It is done and pushed.

File: modules/api/pom.xml (hunk 1 above)
Comment (iizrailevsky) [inline modules/api/pom.xml]: 👍

File: modules/authentication/src/test/java/com/intuit/wasabi/authentication/impl/NoOpAuthenticateByHttpRequestImplTest.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import com.sun.jersey.api.core.HttpRequestContext;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+import static junit.framework.TestCase.assertNull;
+
+public class NoOpAuthenticateByHttpRequestImplTest {
+
+    @Test
+    public void authenticateTest() {
```
Comment (iizrailevsky) [inline modules/authentication/src/test/java/com/intuit/wasabi/authentication/impl/NoOpAuthenticateByHttpRequestImplTest.java]: Awesome - thanks for adding the unit test! 👍
